File: bindings/python/ailoy/vector_store.py
from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, TypeAdapter

from ailoy.runtime import Runtime, generate_uuid


class VectorStoreBaseConfig(BaseModel):
    embedding: Literal["bge-m3"] = "bge-m3"


class FAISSConfig(VectorStoreBaseConfig):
    pass


class ChromadbConfig(VectorStoreBaseConfig):
    url: str
    collection: Optional[str] = None


VectorStoreConfig = Union[FAISSConfig, ChromadbConfig]


class VectorStoreInsertItem(BaseModel):
    document: str
    metadata: Optional[Dict[str, Any]] = None


class VectorStoreRetrieveItem(VectorStoreInsertItem):
    id: str
    similarity: float


class VectorStore:
    """
    The `VectorStore` class provides a high-level abstraction for storing and retrieving documents.
    It mainly consists of two modules - embedding model and vector store.

    It supports embedding text using AI and interfacing with pluggable vector store backends such as FAISS or ChromaDB.
    This class handles initialization, insertion, similarity-based retrieval, and cleanup.

    Typical usage involves:
    1. Initializing the store via `initialize()`
    2. Inserting documents via `insert()`
    3. Querying similar documents via `retrieve()`

    The embedding model and vector store are defined dynamically within the provided runtime.
    """

    def __init__(
        self,
        runtime: Runtime,
        config: VectorStoreConfig = FAISSConfig(),
    ):
        """
        Creates an instance.

        :param runtime: The runtime environment used to manage components.
        :param config: Configuration for the vector store, either FAISSConfig or ChromadbConfig.
        """
        self._runtime = runtime
        self._config = config
        self._vector_store_component_name = generate_uuid()
        self._embedding_model_component_name = generate_uuid()
        self._initialized = False

    def __del__(self):
        self.deinitialize()

    def __enter__(self):
        self.initialize()
        return self

    def __exit__(self, type, value, traceback):
        self.deinitialize()

    def initialize(self):
        """
        Initializes the embedding model and vector store components.
        This must be called before using any other method in the class. If already initialized, this is a no-op.
        """
        if self._initialized:
            return

        # Dimension size may be different based on embedding model
        dimension = 1024

        # Initialize embedding model
        if self._config.embedding == "bge-m3":
            dimension = 1024
            self._runtime.define("tvm_embedding_model", self._embedding_model_component_name, {"model": "BAAI/bge-m3"})

        # Initialize vector store
        if isinstance(self._config, FAISSConfig):
            self._runtime.define("faiss_vector_store", self._vector_store_component_name, {"dimension": dimension})
        elif isinstance(self._config, ChromadbConfig):
            self._runtime.define(
                "chromadb_vector_store",
                self._vector_store_component_name,
                {
                    "url": self._config.url,
                    "collection": self._config.collection,
                },
            )
        else:
            raise ValueError(f"Unsupported vector store: {type(self._config)}")

        self._initialized = True

    def deinitialize(self):
        """
        Deinitializes all internal components and releases resources.
        This should be called when the VectorStore is no longer needed. If already deinitialized, this is a no-op.
        """
        if not self._initialized:
            return

        self._runtime.delete(self._embedding_model_component_name)
        self._runtime.delete(self._vector_store_component_name)
        self._initialized = False

    # TODO: add NDArray typing
    def embedding(self, text: str) -> Any:
        """
        Generates an embedding vector for the given input text using the embedding model.

        :param text: Input text to embed.
        :returns: The resulting embedding vector.
        """
        resp = self._runtime.call_method(
            self._embedding_model_component_name,
            "infer",
            {
                "prompt": text,
            },
        )
        return resp["embedding"]

    def insert(self, document: str, metadata: Optional[Dict[str, Any]] = None):
        """
        Inserts a new document into the vector store.

        :param document: The raw text document to insert.
        :param metadata: Metadata records additional information about the document.
        """
        embedding = self.embedding(document)
        self._runtime.call_method(
            self._vector_store_component_name,
            "insert",
            {
                "embedding": embedding,
                "document": document,
                "metadata": metadata,
            },
        )

    def retrieve(self, query: str, top_k: int = 5) -> List[VectorStoreRetrieveItem]:
        """
        Retrieves the top-K most similar documents to the given query.

        :param query: The input query string to search for similar content.
        :param top_k: Number of top similar documents to retrieve.
        :returns: A list of retrieved items.
        """
        embedding = self.embedding(query)
        resp = self._runtime.call_method(
            self._vector_store_component_name,
            "retrieve",
            {
                "query_embedding": embedding,
                "k": top_k,
            },
        )
        results = TypeAdapter(List[VectorStoreRetrieveItem]).validate_python(resp["results"])
        return results

    def clean(self):
        """
        Removes all entries from the vector store.
        """
        self._runtime.call_method(self._vector_store_component_name, "clean", {})

File: bindings/js-node/src/vector_store.ts
import type { NDArray } from "ailoy_addon.node";

import { Runtime, generateUUID } from "./runtime";

export type EmbeddingModelName = "bge-m3";

export type VectorStoreName = "faiss" | "chromadb";

interface EmbeddingModelDescription {
  modelId: string;
  componentType: "tvm_embedding_model";
  dimension?: number;
}

const modelDescriptions: Record<EmbeddingModelName, EmbeddingModelDescription> =
  {
    "bge-m3": {
      modelId: "BAAI/bge-m3",
      componentType: "tvm_embedding_model",
      dimension: 1024,
    },
  };

export interface VectorStoreInsertItem {
  /** The raw text document to insert */
  document: string;
  /** Metadata records additional information about the document */
  metadata?: Record<string, any>;
}

export interface VectorStoreRetrieveItem extends VectorStoreInsertItem {
  id: string;
  similarity: number;
}

/**
 * The `VectorStore` class provides a high-level abstraction for storing and retrieving documents.
 * It mainly consists of two modules - embedding model and vector store.
 *
 * It supports embedding text using AI and interfacing with pluggable vector store backends such as FAISS or ChromaDB.
 * This class handles initialization, insertion, similarity-based retrieval, and cleanup.
 *
 * Typical usage involves:
 *   1. Initializing the store via `initialize()`
 *   2. Inserting documents via `insert()`
 *   3. Querying similar documents via `retrieve()`
 *
 * The embedding model and vector store are defined dynamically within the provided runtime.
 */
export class VectorStore {
  private runtime: Runtime;

  private componentState: {
    embeddingName: string;
    vecstoreName: string;
    valid: boolean;
  };

  constructor(runtime: Runtime) {
    this.runtime = runtime;

    this.componentState = {
      embeddingName: generateUUID(),
      vecstoreName: generateUUID(),
      valid: false,
    };
  }

  /**
   * Defines the embedding model and vector store components to the runtime.
   * This must be called before using any other method in the class. If already defined, this is a no-op.
   */
  async define(
    embeddingModelName: EmbeddingModelName,
    embeddingModelAttrs: Record<string, any>,
    vectorStoreName: VectorStoreName,
    vectorStoreAttrs: Record<string, any>
  ): Promise<void> {
    // Skip if the component already exists
    if (this.componentState.valid) return;

    if (!(embeddingModelName in modelDescriptions))
      throw Error(`Model "${embeddingModelName}" is not available`);
    const modelDesc = modelDescriptions[embeddingModelName];

    // Add model name into attrs
    if (!embeddingModelAttrs.model)
      embeddingModelAttrs.model = modelDesc.modelId;

    // Initialize embedding model
    const result1 = await this.runtime.define(
      modelDesc.componentType,
      this.componentState.embeddingName,
      embeddingModelAttrs
    );
    if (!result1) throw Error("Embedding model define failed");

    // Add dimension into attrs
    if (!vectorStoreAttrs.dimension && modelDesc.dimension)
      vectorStoreAttrs.dimension = modelDesc.dimension;

    // Initialize vector store
    const vsComponentType = {
      faiss: "faiss_vector_store",
      chromadb: "chromadb_vector_store",
    }[vectorStoreName];
    const result2 = await this.runtime.define(
      vsComponentType,
      this.componentState.vecstoreName,
      vectorStoreAttrs
    );
    if (!result2) throw Error("Embedding model define failed");

    this.componentState.valid = true;
  }

  /**
   * Delete resources from the runtime.
   * This should be called when the VectorStore is no longer needed. If already deleted, this is a no-op.
   */
  async delete(): Promise<void> {
    if (!this.componentState.valid) return;
    await this.runtime.delete(this.componentState.embeddingName);
    await this.runtime.delete(this.componentState.vecstoreName);
    this.componentState.valid = false;
  }

  /** Inserts a new document into the vector store */
  async insert(item: VectorStoreInsertItem): Promise<void> {
    const embedding = await this.embedding(item.document);
    await this.runtime.callMethod(this.componentState.vecstoreName, "insert", {
      embedding: embedding,
      document: item.document,
      metadata: item.metadata,
    });
  }

  /** Retrieves the top-K most similar documents to the given query */
  async retrieve(
    /** The input query string to search for similar content */
    query: string,
    /** Number of top similar documents to retrieve */
    topK: number = 5
  ): Promise<Array<VectorStoreRetrieveItem>> {
    const embedding = await this.embedding(query);
    const resp: { results: Array<VectorStoreRetrieveItem> } =
      await this.runtime.callMethod(
        this.componentState.vecstoreName,
        "retrieve",
        {
          query_embedding: embedding,
          top_k: topK,
        }
      );
    return resp.results;
  }

  async clear(): Promise<void> {
    await this.runtime.callMethod(this.componentState.vecstoreName, "clear");
  }

  /** Generates an embedding vector for the given input text using the embedding model */
  async embedding(
    /** Input text to embed */
    text: string
  ): Promise<NDArray> {
    const resp: { embedding: NDArray } = await this.runtime.callMethod(
      this.componentState.embeddingName,
      "infer",
      {
        prompt: text,
      }
    );
    return resp.embedding;
  }
}

export async function defineVectorStore(
  runtime: Runtime,
  modelName: EmbeddingModelName,
  vectorStoreName: VectorStoreName,
  args?: {
    device?: number;
    url?: string;
    collection?: string;
  }
): Promise<VectorStore> {
  const vs = new VectorStore(runtime);

  const args_ = args || {};

  // Attribute input for call `rt.define(embedding_model)`
  let embeddingAttrs: Record<string, any> = {};
  if (args_.device) embeddingAttrs["device"] = args_.device;

  // Attribute input for call `rt.define(vector_store)`
  let vectorStoreAttrs: Record<string, any> = {};
  if (args_.url) vectorStoreAttrs["url"] = args_.url;
  if (args_.collection) vectorStoreAttrs["collection"] = args_.collection;

  await vs.define(modelName, embeddingAttrs, vectorStoreName, vectorStoreAttrs);

  return vs;
}
